language/vm/vm_runtime/src/block_processor.rs

```diff
@@ -119,19 +123,30 @@ where
     let arena = Arena::new();
     let process_txn = ProcessTransaction::new(txn, &module_cache, data_cache, &arena);
 
-    let validated_txn = match process_txn.validate(mode, publishing_option) {
+    let validated_txn = record_stats! {TXN_VALIDATION_TIME_TAKEN_HISTOGRAM, {
+    match process_txn.validate(mode, publishing_option) {
         Ok(validated_txn) => validated_txn,
         Err(vm_status) => {
             return ExecutedTransaction::discard_error_output(vm_status);
         }
+    }
+    }
     };
-    let verified_txn = match validated_txn.verify(script_cache) {
+
+    let verified_txn = record_stats! {TXN_VERIFICATION_TIME_TAKEN_HISTOGRAM, {
+     match validated_txn.verify(script_cache) {
```

Comment:
keep in mind that cached scripts are not verified again so I am not sure this is going to be very helpful

Reply:
> I think this is good and I like the infra for the counters which we can reuse elsewhere.
> Overall I question most of those counters and not sure I follow the idea.

The idea is to instrument the main "sections" of the transaction execution flow in order to pair this up with the gas usage for that part of the transaction flow e.g. 
* pair-up verification time with the gas amount that we charge for verifying the transaction
* pair-up prologue execution time, and the amount of gas that we will charge for the prologue
etc.
We don't have counters for all of these since e.g. the prologue gas will be defined as a constant. 

> I am unconvinced of time counters based on Instant though that may be a good start or a good way to gather some info for now.

I'm not sure if I quite follow this. What in particular are you worried about w.r.t. this? That they're too fine- or coarse-grained? I replicated here similar to what I saw being done in consensus. We could use the builtin timing mechanism in prometheus, but I have a feeling that will fall back on Instant as well eventually. 

> I am not sure what time based counters are going to do if they don't relate to the same transactions. Prologues and epilogue may be fine as they are stable code but I am not sure what reporting the time on a random transaction is going to tell us.

In their own right, you're totally correct. However, I'm interested in the pairing of these two data dimensions together. One thing we could also do, would be to also bucket by the transaction hash (and this would give us per-transaction-type information). But that isn't as useful (at least to me) right now.

> I put a comment about verification as that is particularly noisy.

Good shout on this. I'll take a look into it.

> General gas usage without a reference to the transaction executed is also not too clear to me.

This ties in with the discussion above -- we aren't concerned about the per-transaction-type gas usage so much as measuring the time taken for a (generic) transaction (or even block of transactions!), and the amount of gas charged for that transaction as well. 

> I am fine going in and the macro stuff is pretty helpful to set it on/off and again this will let us play with the dashboard a bit.

Totally agreed, I think this is something that we should look at more in terms of instrumenting things, macros can be very handy for this since it'll allow us to compile these things away.